### PR TITLE
feat: Skip sampler for intermediate pp stages.

### PR DESCRIPTION
**Pull request by yuxianq**

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 1]
```diff
@@ -81,13 +82,14 @@ def update_requests(self, state: SampleState) -> None:
...
             # NOTE: This is a hack: set finish reason manually and set the beam 0
             request.set_finished_reason(FinishReason.LENGTH, 0)
-            logits = state.logits[idx]
-            if logits.ndim == 1:
-                # For BERT: Add axis to be compatible with LogitsStorage
-                # (LogitsStorage will interpret this dim as the prompt_len which
-                # is not relevant for outputting logits of encoder only model).
-                logits = logits.unsqueeze(0)
-            request.py_result.append_context_logits(logits)
+            if request.py_return_context_logits:
```
Comment (yuxianq) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:85]: @symphonylyh Please help to review whether we get logits only when `request.py_return_context_logits` is True.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 2]
```diff
@@ -656,6 +654,15 @@ def _executor_loop_cleanup(self):
             self.response_cv.notify_all()
         self.shutdown_event.set()
 
+    def _need_return_logits(self, scheduled_requests: ScheduledRequests):
```
Comment (yuxianq) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py:657]: @tongyuantongyu Please help to review whether this `_need_return_logits` is correct. Do we only need logits when there is `py_return_generation_logits/py_return_generation_logits = True`?

File: tensorrt_llm/_torch/pyexecutor/sampler.py (hunk 1 above)
Comment (yuxianq) [reply tensorrt_llm/_torch/pyexecutor/sampler.py:85]: This condition is for skipping `logits = state.logits[idx]` since we won't send logits as part of sampler state from the last pp stage if `_need_return_logits` returns False, which means `request.py_return_context_logits` is False. `state.logits` will be None in this case.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py (hunk 2 above)
Comment (yuxianq) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py:657]: Got it. Sometimes log_probs has been calculated in `sample_async`, so we directly send log_probs instead of logits to other pp stages in this case, which is much faster. We can detect this case by checking whether log_probs is None in `SampleState`. Do you know how to detect whether we need logits to compute log_probs in `update_requests`? Should we always send logits in this case?

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 3]
```diff
@@ -737,33 +758,51 @@ def _executor_loop_pp(self):
...
                         if self.send_handles[prev_microbatch_id] is not None:
                             self.send_handles[prev_microbatch_id].Wait()
                         self.send_handles[
-                            prev_microbatch_id] = self.dist.isend_tensor_list(
-                                previous_batch.sample_state.host.values(),
+                            prev_microbatch_id] = self.dist.isend_object(
+                                (
+                                    sample_state.logits_host.numpy() if
+                                    self._need_return_logits(scheduled_batch) or
+                                    (self._need_return_log_probs(
```
Comment (yuxianq) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py:799]: @tongyuantongyu I add check for `py_return_context_logits` when sending logits according to your advice.

**Review by amukkara — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 4]
```diff
@@ -737,33 +758,51 @@ def _executor_loop_pp(self):
...
+                            sample_state.new_tensors_host,
+                        ) = self.dist.recv_object(
                             src=self.dist.prev_pp_rank,
-                            tag=prev_microbatch_id)
+                            tag=prev_microbatch_id,
+                        )
+                        if logits is not None:
+                            logits_host = torch.from_numpy(logits)
+                            sample_state.logits_host = logits_host
+                            sample_state.logits = logits_host.to(self.device_id)
```
[the remark this answers is not in the document]
Comment (yuxianq) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py:784]: We should blocking memcpy it since we will send it to the next pp stage in the current step immediately in L795: https://github.com/NVIDIA/TensorRT-LLM/blob/6bdecbb0905716c71834287c480ac409232ff185/tensorrt_llm/_torch/pyexecutor/py_executor.py#L795

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 5]
```diff
@@ -656,6 +654,24 @@ def _executor_loop_cleanup(self):
             self.response_cv.notify_all()
         self.shutdown_event.set()
 
+    def _need_return_logits(self, scheduled_requests: ScheduledRequests):
+        for req in scheduled_requests.context_requests:
+            if req.py_return_generation_logits:
```
[the remark this answers is not in the document]
Comment (yuxianq) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: Nice catch, fixed.

**Comment by yuxianq** [comment]
/bot run --disable-fail-fast --stage-list "DGX_H100-4_GPUs-PyTorch-DeepSeek-1, DGX_H100-4_GPUs-PyTorch-Others-1"

**Comment by yuxianq** [comment]
/bot run --disable-fail-fast --stage-list "DGX_H100-4_GPUs-PyTorch-DeepSeek-1, DGX_H100-4_GPUs-PyTorch-Others-1"

**Comment by yuxianq** [comment]
/bot run --disable-fail-fast --stage-list "DGX_H100-4_GPUs-PyTorch-DeepSeek-1, DGX_H100-4_GPUs-PyTorch-Others-1"

**Comment by yuxianq** [comment]
/bot run --disable-fail-fast --stage-list "DGX_H100-4_GPUs-PyTorch-DeepSeek-1, DGX_H100-4_GPUs-PyTorch-Others-1"

**Comment by yuxianq** [comment]
/bot skip --comment "The failed test unittest/llmapi/test_llm_multi_gpu.py::test_llm_multi_node_with_postproc is unrelated to this PR and caused by https://github.com/NVIDIA/TensorRT-LLM/pull/3869, all of other tests have been passed."

**Review by yizhang-nv — APPROVED** [review]

**Review by litaotju — APPROVED** [review]
